computation_graph/graph.py

```diff
@@ -15,28 +16,48 @@ def _is_reducer_type(node: Callable) -> bool:
     return "state" in inspect.signature(node).parameters
 
 
-def infer_callable_signature(func: Callable) -> base_types.NodeSignature:
+def _is_star(parameter) -> bool:
+    return "*" + parameter.name == str(parameter)
+
+
+def _is_double_star(parameter) -> bool:
+    return "**" + parameter.name == str(parameter)
+
+
+def _is_default(parameter):
+    return parameter.default != parameter.empty
+
+
+_parameter_name = gamla.attrgetter("name")
+
+
+@gamla.before(
+    gamla.compose_left(
+        inspect.signature,
+        gamla.attrgetter("parameters"),
+        MappingProxyType.values,
+        tuple,
+    )
+)
+def _infer_callable_signature(function_parameters: Tuple) -> base_types.NodeSignature:
     signature = base_types.NodeSignature(
-        is_args=any(
-            "*" + x.name == str(x)
-            for x in tuple(inspect.signature(func).parameters.values())
+        is_args=gamla.anymap(_is_star)(function_parameters),
+        kwargs=gamla.pipe(
+            function_parameters,
+            gamla.remove(_is_star),
+            gamla.map(_parameter_name),
+            tuple,
         ),
-        kwargs=tuple(
-            x.name
-            for x in tuple(inspect.signature(func).parameters.values())
-            if "*" + x.name != str(x)
-        ),
-        optional_kwargs=tuple(
-            x.name
-            for x in tuple(inspect.signature(func).parameters.values())
-            if "*" + x.name != str(x) and x.default != x.empty
+        optional_kwargs=gamla.pipe(
+            function_parameters,
+            gamla.filter(gamla.alljuxt(gamla.complement(_is_star), _is_default)),
+            gamla.map(_parameter_name),
+            tuple,
         ),
     )
 
-    assert not any(
-        "**" + x.name == str(x)
-        for x in tuple(inspect.signature(func).parameters.values())
-    ), f"Cannot infer signature for {func}, signature includes **kwargs"
+    if gamla.anymap(_is_double_star)(function_parameters):
```

Comment:
move this to the top of the function and then you can remove the `signature` param